### ISSUE TITLE
[WARNING] The POM for org.apache.jmeter:ApacheJMeter_http:jar:5.6.1 is invalid

### ISSUE THREAD

**Issue by jloisel** [apache/jmeter#6041]
### Expected behavior

When trying to depend on Apache JMeter http dependency:

> [WARNING] The POM for org.apache.jmeter:ApacheJMeter_http:jar:5.6.1 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details

> java: package org.apache.jmeter.testelement does not exist

![image](https://github.com/apache/jmeter/assets/3208716/ace269a3-9f7a-4a92-b82c-865cc05a8381)

### Steps to reproduce the problem

```xml
  <dependencies>
    <dependency>
      <groupId>org.apache.jmeter</groupId>
      <artifactId>ApacheJMeter_http</artifactId>
      <version>5.6.1</version>
    </dependency>
  </dependencies>
```

### JMeter Version

5.6.1

### Java Version

openjdk version "17.0.7" 2023-04-18

### OS Version

Ubuntu 22.04 LTS

**Comment by vlsi** [comment]
```
[WARNING] The POM for org.apache.jmeter:ApacheJMeter_http:jar:5.6.1 is invalid, transitive dependencies (if any) will not be available: 1 problem was encountered while building the effective model for org.apache.jmeter:ApacheJMeter_http:5.6.1
[ERROR] 'dependencies.dependency.version' for com.google.auto.service:auto-service-annotations:jar is missing. @
```

**Comment by vlsi** [comment]
Frankly speaking, I agree it might look confusing for a pom to miss `<version>`, however, a missing version is in line with XSD.
So I think it would make sense to follow-up with Maven developers, and request that Maven should not discard all the transitive dependencies once it faces a single one without a version.

**Comment by milamberspace** [comment]
org.apache.jmeter.samplers.SampleListener is in ApacheJMeter_core not _http?

**Comment by milamberspace** [comment]
Works with this pom

<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>org.example</groupId>
    <artifactId>TestBugMaven</artifactId>
    <version>1.0-SNAPSHOT</version>

    <properties>
        <maven.compiler.source>17</maven.compiler.source>
        <maven.compiler.target>17</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>

    <dependencies>
        <dependency>
            <groupId>org.apache.jmeter</groupId>
            <artifactId>ApacheJMeter_core</artifactId>
            <version>5.6.1</version>
        </dependency>
    </dependencies>
</project>

**Comment by vlsi** [comment]
@milamberspace , try running `mvn dependency:tree` with that pom. it will not show transitive dependencies.

```
[WARNING] The POM for org.apache.jmeter:ApacheJMeter_core:jar:5.6.1 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
[INFO]
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ TestBugMaven ---
[INFO] org.example:TestBugMaven:jar:1.0-SNAPSHOT
[INFO] \- org.apache.jmeter:ApacheJMeter_core:jar:5.6.1:compile
```

Then, if you patch `~/.m2/repository/org/apache/jmeter/ApacheJMeter_core/5.6.1` by adding `<version>1.1.1</version>` to `auto-service-annotations`, you'll get a dramatic difference:

```
[INFO] --- maven-dependency-plugin:2.8:tree (default-cli) @ TestBugMaven ---
[INFO] org.example:TestBugMaven:jar:1.0-SNAPSHOT
[INFO] \- org.apache.jmeter:ApacheJMeter_core:jar:5.6.1:compile
[INFO]    +- org.apache.jmeter:ApacheJMeter:jar:5.6.1:compile
[INFO]    +- org.apache.jmeter:jorphan:jar:5.6.1:compile
[INFO]    +- bsf:bsf:jar:2.4.0:compile
[INFO]    |  \- commons-logging:commons-logging:jar:1.0.4:compile
[INFO]    +- com.fifesoft:rsyntaxtextarea:jar:3.3.3:compile
[INFO]    +- net.sf.jtidy:jtidy:jar:r938:compile
[INFO]    +- com.thoughtworks.xstream:xstream:jar:1.4.20:compile
[INFO]    |  \- io.github.x-stream:mxparser:jar:1.2.2:compile
[INFO]    |     \- xmlpull:xmlpull:jar:1.1.3.1:compile
[INFO]    +- org.apache.logging.log4j:log4j-1.2-api:jar:2.20.0:compile
[INFO]    +- org.apache.logging.log4j:log4j-api:jar:2.20.0:compile
[INFO]    +- org.apache.logging.log4j:log4j-core:jar:2.20.0:compile
[INFO]    +- org.apache.logging.log4j:log4j-slf4j-impl:jar:2.20.0:compile
[INFO]    |  \- org.slf4j:slf4j-api:jar:1.7.25:compile
[INFO]    +- org.apiguardian:apiguardian-api:jar:1.1.2:compile
[INFO]    +- oro:oro:jar:2.0.8:compile
[INFO]    +- xalan:xalan:jar:2.7.3:compile
[INFO]    +- xalan:serializer:jar:2.7.3:compile
[INFO]    +- xml-apis:xml-apis:jar:1.4.01:compile
[INFO]    +- net.sf.saxon:Saxon-HE:jar:11.5:compile
[INFO]    |  +- org.xmlresolver:xmlresolver:jar:4.6.4:compile
[INFO]    |  |  +- org.apache.httpcomponents.client5:httpclient5:jar:5.1.3:runtime
[INFO]    |  |  |  \- org.apache.httpcomponents.core5:httpcore5-h2:jar:5.1.3:runtime
[INFO]    |  |  \- org.apache.httpcomponents.core5:httpcore5:jar:5.1.3:runtime
[INFO]    |  \- org.xmlresolver:xmlresolver:jar:data:4.6.4:compile
[INFO]    +- org.jetbrains.kotlin:kotlin-stdlib-jdk8:jar:1.8.22:compile
[INFO]    |  +- org.jetbrains.kotlin:kotlin-stdlib:jar:1.8.22:compile
[INFO]    |  \- org.jetbrains.kotlin:kotlin-stdlib-jdk7:jar:1.8.22:compile
[INFO]    +- com.google.auto.service:auto-service-annotations:jar:1.1.1:compile
[INFO]    +- org.jetbrains.kotlinx:kotlinx-coroutines-core:jar:1.7.1:runtime
...
```

Apparently, Maven discards ALL transitive dependency in case it detects a single issue. I think it is something to be fixed in Maven. In this case, only `com.google.auto.service:auto-service-annotations` was missing a version, and it is not essential overall. If Maven discarded a single problematic dependency or failed the build, then it would be way better.

**Comment by vlsi** [comment]
I've filed an issue for the Maven team: https://issues.apache.org/jira/browse/MNG-7842
I believe Maven's behavior is not really consistent as `dependencyManagement` does not help.

**Comment by Ardesco** [comment]
I would suggest this still needs a fix.  Even if Maven does fix this issue it would require all Maven users to upgrade to the latest version of Maven which can be problematic for many corp environments.

For the sake of backwards compatibility I think it would make sense to release a 5.6.2 with the version number specified.  You can then deprecate old versions of Maven supported in the future if/when Maven changes the behaviour and start not specifying versions at that point.

Personally I would suggest it's good practice to specify versions, overall it makes it clearer.

**Comment by Ardesco** [comment]
> Works with this pom
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <project xmlns="http://maven.apache.org/POM/4.0.0"
>          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
>          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
>     <modelVersion>4.0.0</modelVersion>
> 
>     <groupId>org.example</groupId>
>     <artifactId>TestBugMaven</artifactId>
>     <version>1.0-SNAPSHOT</version>
> 
>     <properties>
>         <maven.compiler.source>17</maven.compiler.source>
>         <maven.compiler.target>17</maven.compiler.target>
>         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
>     </properties>
> 
>     <dependencies>
>         <dependency>
>             <groupId>org.apache.jmeter</groupId>
>             <artifactId>ApacheJMeter_core</artifactId>
>             <version>5.6.1</version>
>         </dependency>
>     </dependencies>
> </project>
> ```

This only fixes the problem if JMeter is a direct dependency.  If it's a transitive dependency you are still getting error messages saying that the POM is invalid and that transitive dependencies will not be available.

**Comment by hennidk** [comment]
In my case I am unable to compile my project, getting the error:
```
... 
Compilation failure
[ERROR]   reason: class file for org.apiguardian.api.API$Status not found
[ERROR] warning: unknown enum constant Status.EXPERIMENTAL
...
error: cannot access Function0
  class file for kotlin.jvm.functions.Function0 not found
```
Appears to be the same problem as mentioned where Maven discards everything - surely fixing the **invalid POM** error as suggested is the best way forward?

**Comment by jloisel** [comment]
I also believe versions should always be specified, otherwise it might lead to inconsistencies (which version is it supposed to take then?).

**Comment by vlsi** [comment]
>I also believe versions should always be specified, otherwise it might lead to inconsistencies (which version is it supposed to take then?).

The version might be taken from a `dependencyManagement` block.

**Comment by vlsi** [comment]
This seems to be resolved in 5.6.2-rc1:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>org.example</groupId>
    <artifactId>TestBugMaven</artifactId>
    <version>1.0-SNAPSHOT</version>

    <dependencies>
        <dependency>
            <groupId>org.apache.jmeter</groupId>
            <artifactId>ApacheJMeter_http</artifactId>
            <version>5.6.2</version>
        </dependency>
    </dependencies>

    <repositories>
      <repository>
        <id>jmeter-5.6.2-rc1</id>
        <url>https://repository.apache.org/content/repositories/orgapachejmeter-1087</url>
      </repository>
    </repositories>
</project>
```

Please try it.

**Comment by Ardesco** [comment]
That version looks good to me, thanks for the quick turnaround.
